custom_components/ha_creality_ws/utils.py

```
@@ -330,4 +337,285 @@ def resolved_model(self) -> str:
+    }
+    if not payload["type"]:
+        raise ValueError("material type must not be empty")
+
+    for key, value in (("name", name), ("vendor", vendor)):
+        if value is not None and str(value).strip():
+            payload[key] = str(value).strip()
+
+    if color is not None:
+        payload["color"] = normalize_material_color(color)
+
+    def _number(name: str, value: Any) -> float | None:
```

Ruff (0.16.1)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `value`

(ANN401)